test/e2e/storage/csi_mock_volume.go

```diff
@@ -0,0 +1,468 @@
+				init()
+				defer cleanup()
+
+				_, _, pod := createPod()
+				if pod == nil {
+					return
+				}
+				err = framework.WaitForPodNameRunningInNamespace(m.cs, pod.Name, pod.Namespace)
+				framework.ExpectNoError(err, "Failed to start pod: %v", err)
+				By("Checking CSI driver logs")
+
+				// The driver is deployed as a statefulset with stable pod names
+				driverPodName := "csi-mockplugin-0"
+				err = checkPodInfo(m.cs, f.Namespace.Name, driverPodName, "mock", pod, test.expectPodInfo)
+				framework.ExpectNoError(err)
+			})
+		}
+	})
+
+	Context("CSI volume limit information using mock driver", func() {
+		tests := []struct {
+			name           string
+			csiAttachLimit int
+		}{
+			{
+				name:           "should report attach limit when bigger than 0",
+				csiAttachLimit: 2,
+			},
+		}
+		for _, t := range tests {
+			test := t
+			It(t.name, func() {
+				attachable = true
+				nodeSelectorKey = "node-attach-limit-csi"
+				attachLimit = test.csiAttachLimit
+				var err error
+
+				init()
+				defer cleanup()
+				nodeName := m.config.ClientNodeName
+				attachKey := v1.ResourceName(volumeutil.GetCSIAttachLimitKey(m.provisioner))
+
+				nodeAttachLimit, err := checkNodeForLimits(nodeName, attachKey, m.cs)
+				Expect(err).NotTo(HaveOccurred(), "while fetching node %v", err)
+
+				Expect(nodeAttachLimit).To(Equal(test.csiAttachLimit))
+
+				_, _, pod1 := createPod()
+				Expect(pod1).NotTo(BeNil(), "while creating first pod")
+
+				err = framework.WaitForPodNameRunningInNamespace(m.cs, pod1.Name, pod1.Namespace)
+				framework.ExpectNoError(err, "Failed to start pod1: %v", err)
+
+				_, _, pod2 := createPod()
+				Expect(pod2).NotTo(BeNil(), "while creating second pod")
+
+				err = framework.WaitForPodNameRunningInNamespace(m.cs, pod2.Name, pod2.Namespace)
+				framework.ExpectNoError(err, "Failed to start pod2: %v", err)
+
+				_, _, pod3 := createPod()
+				Expect(pod3).NotTo(BeNil(), "while creating third pod")
+				err = waitForMaxVolumeCondition(pod3, m.cs)
+				Expect(err).NotTo(HaveOccurred(), "while waiting for max volume condition")
+			})
+		}
+	})
+
+})
+
+func waitForMaxVolumeCondition(pod *v1.Pod, cs clientset.Interface) error {
+	var err error
+	waitErr := wait.PollImmediate(10*time.Second, 2*time.Minute, func() (bool, error) {
+		pod, err = cs.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		conditions := pod.Status.Conditions
+		for _, condition := range conditions {
+			matched, _ := regexp.MatchString("max.+volume.+count", condition.Message)
+			if condition.Reason == v1.PodReasonUnschedulable && matched {
+				return true, nil
+			}
+
+		}
+		return false, nil
+	})
+	return waitErr
+}
+
+func checkNodeForLimits(nodeName string, attachKey v1.ResourceName, cs clientset.Interface) (int, error) {
+	var attachLimit int64
+
+	waitErr := wait.PollImmediate(10*time.Second, 5*time.Minute, func() (bool, error) {
```

Comment:
Dtto, It should be a const somewhere (same as any other node condition timeout?)